content/post/method-data-scalability/index.adoc

```diff
@@ -0,0 +1,149 @@
+date: 2025-08-07T00:00:00Z
+categories: ['performance', 'benchmarking', 'methodology']
+summary: 'Understand how Tiered Compilation impacts the scalability of Java applications in modern environments.'
+image: 'sharing_is_scaring.png'
+related: ['']
+authors:
+ - Francesco Nigro
+---
+# JVM Challenges in Containers
+
+Containers have revolutionized software deployment, offering lightweight, portable, and consistent environments. With orchestration platforms like Kubernetes, developers can efficiently deploy and scale applications across diverse infrastructures.
+
+However, containers pose unique challenges for applications with complex runtime requirements, such as those running on the Java Virtual Machine (JVM). The JVM, a cornerstone of enterprise software, was designed in an era when it could assume unrestricted access to system resources. Containers, on the other hand, abstract these resources and often impose limits on CPU, memory, and other critical parameters.
+
+While the JVM has evolved to better handle containerized environments — adding features like container resource detection — some components, like Just-In-Time (JIT) compilers (C1 and C2), remain sensitive to resource constraints. Misconfigurations or insufficient resources can significantly impact their efficiency, affecting application performance.
+
+To achieve optimal JVM performance in containers, developers must understand the underlying system and carefully configure resources. Containers simplify deployment but do not inherently address JVM-specific needs.
+
+This article explores how resource shortages impact Java application performance, focusing on the scalability challenges introduced by Tiered Compilation.
+
+# Understanding Tiered Compilation
+
+First, let’s recall a key mechanism employed by OpenJDK Hotspot to optimize the application’s code: Tiered Compilation.
+
+Tiered compilation in the HotSpot JVM balances application startup speed and runtime performance by using https://developers.redhat.com/articles/2021/06/23/how-jit-compiler-boosts-java-performance-openjdk[multiple levels] of code execution and optimization.
+Initially, it uses an **interpreter** for immediate execution. As methods are invoked frequently, it employs a fast compiler i.e. **C1** to generate native code.
+Over time, methods that are heavily used ("hot spots") are further optimized with the optimizing compiler i.e. **C2**, which applies advanced optimizations for maximum performance.
+
+This tiered approach ensures quick application responsiveness while progressively optimizing performance-critical code paths. The name "HotSpot" reflects this focus on dynamically identifying and optimizing hot spots in code execution for efficiency.
+
+What’s less known about tiered compilation is that the C2 compiler can be very CPU intensive and, when it doesn’t have enough resources, its activity https://jpbempel.github.io/2020/05/22/startup-containers-tieredcompilation.html[affects startup time].
```

Comment:
Just a minor but if I am not wrong the C2 compilation could also affect runtime performance and not only the startup, if so I think it would be good to clarify that here
```suggestion
What’s less known about tiered compilation is that the C2 compiler can be very CPU intensive and, when it doesn’t have enough resources, its activity could affect application performances including https://jpbempel.github.io/2020/05/22/startup-containers-tieredcompilation.html[startup time].
```